[JeremyYong128] iP

JeremyYong128 opened the pull request:
### DukePro

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://piratelol.ytmnd.com))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://piratelol.ytmnd.com).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

JeremyYong128 committed:
Increment: Level-0

JeremyYong128 committed:
Increment: Level-1

JeremyYong128 committed:
Increment: Level-2

JeremyYong128 committed:
Increment: Level-3

JeremyYong128 committed:
Increment: Level-4

JeremyYong128 committed:
Increment: A-TextUiTesting

JeremyYong128 committed:
Increment: A-TextUiTesting

JeremyYong128 committed:
Increment: Level-5

JeremyYong128 committed:
Increment: Level-6

JeremyYong128 committed:
Increment: Level-7

JeremyYong128 committed:
Increment: Level-8

JeremyYong128 committed:
Increment: A-MoreOOP

JeremyYong128 committed:
Increment: A-Packages

JeremyYong128 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-Level-7

JeremyYong128 committed:
Increment: A-Gradle

JeremyYong128 committed:
Increment: A-JUnit

JeremyYong128 committed:
Increment: A-Jar

JeremyYong128 committed:
Increment: A-JavaDoc

JeremyYong128 committed:
Ensure code complies with coding standards.

JeremyYong128 committed:
Add feature to find tasks containing a substring.

JeremyYong128 committed:
Merge remote-tracking branch 'origin/branch-A-CodingStandard'

JeremyYong128 committed:
Merge branch 'branch-Level-9'

Dioclei commented on src/main/java/duke/DateManager.java:
```diff
@@ -0,0 +1,79 @@
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+
+public class DateManager {
```
Perhaps you can add a short header comment at the start of each class to describe what the class does!

Dioclei commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @return String representation of the Event in the storage.
+     */
+    public String toStringStorage() {
+        String nameField = this.getName();
+        String isDoneField = this.isDone() ? "1" : "0";
+        String fromField = this.from;
+        String toField = this.to;
+        return "E|" + isDoneField + "|" + nameField + "|" + fromField + "|" + toField;
+    }
+}
```
Well done, this piece of code is really readable.

Dioclei commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+                        } catch (NumberFormatException e) {
+                            throw new WrongInputException();
+                        }
+                    }
+                } else if (splitStr[0].equals("todo") || splitStr[0].equals("deadline") || splitStr[0].equals("event")) {
+                    Task addedTask = new Task("");
+                    switch (splitStr[0]) {
+                    case "todo":
+                        String[] todoSplit = s.split("todo");
+                        if (todoSplit.length <= 1 || todoSplit[1].trim().equals("")) {
+                            throw new EmptyTodoException();
+                        }
+                        addedTask = new Todo(todoSplit[1].trim());
+                        break;
+                    case "deadline":
+                        String splitStr1 = s.split("deadline ")[1];
+                        String[] splitStr2 = splitStr1.split(" /by ");
+                        String[] splitStr3 = splitStr2[1].split(" ");
+                        LocalDateTime dateTime = null;
+                        if (splitStr3.length == 2) {
+                            dateTime = DateManager.parseDateString(splitStr3[0], splitStr3[1]);
+                        }
+                        addedTask = dateTime == null ? new Deadline(splitStr2[0], splitStr2[1])
+                                : new Deadline(splitStr2[0], dateTime);
+                        break;
+                    case "event":
+                        String eSplitStr1 = s.split("event ")[1];
+                        String[] eSplitStr2 = eSplitStr1.split(" /");
+                        addedTask = new Event(eSplitStr2[0], eSplitStr2[1].substring(5), eSplitStr2[2].substring(3));
+                        break;
+                    }
+                    tasks.addTask(addedTask);
+                    Ui.informTaskAdded(addedTask, tasks.getSize());
+                } else if (splitStr[0].equals("find") && splitStr.length > 1 && !splitStr[1].equals("")) {
+                    String subString = s.substring(5);
+                    ArrayList<Task> filteredList = tasks.filterBySubstring(subString);
+                    Tasklist newTasklist = new Tasklist(filteredList);
+                    Ui.listTasks(newTasklist);
+                } else {
+                    throw new WrongInputException();
```
I feel that this part is nested very deeply, with try-catch blocks nested inside each other, and this makes the code hard to read. Perhaps you can make it one big try-catch block and remove the inner nested ones. You can also break down the method into a few sub methods to make it more readable.

Dioclei commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,105 @@
+                }
+
+                String identifier = parts[0];
+                boolean done = parts[1] == "1";
```
I feel that this boolean should be named "isDone" to reflect that it is a boolean.

Dioclei reviewed (commented):
Great PR! Code is generally well written but some parts can be broken down to improve readability. 👍 👍

JeremyYong128 committed:
Add GUI.

Sasmik23 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.ArrayList;
+
+public class Parser {
+    /**
+     * Parses the response given by the user.
+     *
+     * @param s The response of the user.
+     * @param tasks The current Tasklist.
+     */
```
Javadocs can be updated to include the last param and return type!

Sasmik23 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,92 @@
+                        if (splitStr3.length == 2) {
+                            dateTime = DateManager.parseDateString(splitStr3[0], splitStr3[1]);
+                        }
+                        addedTask = dateTime == null ? new Deadline(splitStr2[0], splitStr2[1])
+                                : new Deadline(splitStr2[0], dateTime);
```
from the 2103 textbook! :

Here are two acceptable ways to format ternary expressions:
alpha = (aLongBooleanExpression) ? beta : gamma;
alpha = (aLongBooleanExpression)
&emsp;&emsp;&emsp;&emsp;? beta
&emsp;&emsp;&emsp;&emsp;: gamma;

Sasmik23 reviewed (commented):
Overall easy to follow code with Javadocs given for most functions. Might want to avoid too much deep nesting in the parser file, perhaps could abstract out certain lines to the task class? Otherwise good job!

JeremyYong128 committed:
Assumptions at certain stages of the program are not verified when
calling certain methods.

If the assumptions do not hold, the program might throw runtime errors
when a user is interacting with it.

Add assertions to determine if these assumptions hold at point of
execution of the methods.

These will highlight any cases in which the assumptions do not hold so
that attention can be drawn to bugs in the code.

JeremyYong128 committed:
Change the build.gradle file to allow Gradle to build the jar file.

JeremyYong128 committed:
Improve code quality.

Code is messy in some areas. These include:
* Chunks of commented code
* Redundant initialisation of variables
* Unused methods
* Unnecessary nesting

Clean up these areas to improve readability of code.

JeremyYong128 committed:
Add feature to detect duplicates.

Users are able to add multiple tasks which have the same attributes.

This is redundant as having 2 identical tasks only makes the list harder
to read.

Upon adding a new task, the bot will not add the task if it already
exists and inform the user of an existing task.

This prevents duplicates from being added to the list.

JeremyYong128 committed:
Merge pull request #1 from JeremyYong128/branch-A-Assertions

Merge assertions

JeremyYong128 committed:
Merge branch 'master' into branch-A-CodeQuality

JeremyYong128 committed:
Merge pull request #2 from JeremyYong128/branch-A-CodeQuality

Improve code quality

JeremyYong128 committed:
Merge branch 'master' into branch-C-DetectDuplicates

JeremyYong128 committed:
Merge pull request #3 from JeremyYong128/branch-C-DetectDuplicates

Add feature to detect duplicates.

JeremyYong128 committed:
Fix merge conflict issues

Pull request merge conflicts were resolved wrongly previously, removing
the functionality of the Parser to check for duplicates.

Add the functionality into the Parser class to ensure duplicate
detection works.

JeremyYong128 committed:
Add user guide.

Add user documentation for users to understand the functionality of the
chatbot.
